lakeroad-egglog/egglog_src/lakeroad.egg

```diff
@@ -0,0 +1,190 @@
+;;; Lakeroad language definition, for importing before running other tests.
+
+; Ops
+(datatype Op
+  (And)
+  (Or)
+  (Xor)
+  (Shr)
+  ; Returns a bitvector of width 1.
+  (Eq)
+  (LogicNot)
+  ; (Mux select-expr expr expr)
+  (Mux))
+
+; Sketch Ops
+(datatype SketchOp
+  (LUT4))
+
+; Our language. Note that we do not explicitly use IDs as in the Lakeroad paper
+; formalization. The IDs are still there, though: they're the 
+; eclass IDs!
+(datatype Expr
+  ; (BV value bitwidth)
+  (BV i64 i64)
+  ; (Var name bitwidth)
+  (Var String i64)
+  ; (OpN op input-expr...)
+  (Op1 Op Expr :cost 100000000)
+  (Op2 Op Expr Expr :cost 100000000)
+  (Op3 Op Expr Expr Expr :cost 100000000)
+  ; (Reg default-value clock-expr d-expr)
+  (Reg i64 Expr Expr)
+  ; (Wire name bitwidth)
+  ; Wire is a placeholder for another expression. See below how 
+  ; wires are the key to creating cyclic graphs.
+  (Wire String i64)
+  ; (Extract high low expr)
+  ; Extracts expr[high:low] using inclusive ranges as in Rosette/Verilog.
+  (Extract i64 i64 Expr)
+  ; (Concat expr expr)
+  (Concat Expr Expr)
+  ; (ZeroExtend expr bitwidth)
+  (ZeroExtend Expr i64)
+  
+  ; (Sketch sketch-id input-expr...)
+  ; A sketch node. This is a proposal for a potential expression, that will
+  ; need to be solved for/made concrete later on.
+  (Sketch1 SketchOp Expr)
+  (Sketch2 SketchOp Expr Expr)
+  (Sketch3 SketchOp Expr Expr Expr)
+  (Sketch4 SketchOp Expr Expr Expr Expr)
+  )
+
+;;; Types for Lakeroad expressions.
+(datatype Type
+  ;;; Bitvector type.
+  (Bitvector i64))
+
+;;; Indicates that a Lakeroad expression has a given type.
+(relation HasType (Expr Type))
+
+;;; Typing judgements.
+(ruleset typing)
+(rule
```

Comment:
Hi, you need to add an additional rule here if you want things to type properly:
```
(rule
  ((Wire name bw))
  ((HasType (Wire name bw) (Bitvector bw)))
  :ruleset typing)
```

Otherwise, wires that are `Op`ed together will have the wrong type. You also need to saturate the typing rule BEFORE you delete the wires for this to work. Here's a small example:
```
(include "common/lakeroad_rewrites.egg")

; wire declarations
; $and$half-adder.v:17$2_Y
(let v0 (Wire "v0" 1))
; $xor$half-adder.v:16$1_Y
(let v1 (Wire "v1" 1))
; i_bit2
(let v2 (Wire "v2" 1))
; i_bit1
(let v3 (Wire "v3" 1))

; cells
(union v0 (Op2 (And) v3 v2))


(union v1 (Op2 (Xor) v3 v2))

; inputs
(let i_bit2 (Var "i_bit2" 1))
(union v2 i_bit2)
(let i_bit1 (Var "i_bit1" 1))
(union v3 i_bit1)

; outputs
(let o_carry v0)
(let o_sum v1)

; delete wire expressions

(run-schedule
 (repeat 100
         (saturate typing)
         ;; (saturate sketch-proposal)
         ;; (saturate misc)
         ))

(delete (Wire "v0" 1))
(delete (Wire "v1" 1))
(delete (Wire "v2" 1))
(delete (Wire "v3" 1))

(run-schedule
 (repeat 100
         (saturate typing)
         ;; (saturate sketch-proposal)
         ;; (saturate misc)
         ))

(extract o_sum)

(query-extract (HasType o_sum (Bitvector 1)))
```
Thanks.